presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsTracker.java

```diff
@@ -137,17 +142,46 @@ public Map<PlanNodeWithHash, PlanStatisticsWithSourceInfo> getQueryStats(QueryIn
             boolean isScaledWriterStage = stageInfo.getPlan().isPresent() && stageInfo.getPlan().get().getPartitioning().equals(SCALED_WRITER_DISTRIBUTION);
             PlanNode root = stageInfo.getPlan().get().getRoot();
             for (PlanNode planNode : forTree(PlanNode::getSources).depthFirstPreOrder(root)) {
-                if (!planNode.getStatsEquivalentPlanNode().isPresent()) {
+                if (!planNode.getStatsEquivalentPlanNode().isPresent() && !isAggregation(planNode, AggregationNode.Step.PARTIAL)) {
                     continue;
                 }
                 PlanNodeStats planNodeStats = planNodeStatsMap.get(planNode.getId());
                 if (planNodeStats == null) {
                     continue;
                 }
+
                 double outputPositions = planNodeStats.getPlanNodeOutputPositions();
                 double outputBytes = adjustedOutputBytes(planNode, planNodeStats);
                 double nullJoinBuildKeyCount = planNodeStats.getPlanNodeNullJoinBuildKeyCount();
                 double joinBuildKeyCount = planNodeStats.getPlanNodeJoinBuildKeyCount();
+                AggregationNodeStatistics partialAggregationStatistics = AggregationNodeStatistics.empty();
+
+                if (isAggregation(planNode, AggregationNode.Step.PARTIAL)) {
+                    // update stats of corresponding Final Aggregation
+                    partialAggregationStatistics = constructAggregationNodeStatistics(planNode, planNodeStatsMap, outputBytes);
+
+                    List<VariableReferenceExpression> groupingCols = ((AggregationNode) planNode).getGroupingKeys();
+                    for (Map.Entry<PlanNodeWithHash, PlanStatisticsWithSourceInfo> mapEntry : aggregationNodeStats.entrySet()) {
+                        AggregationNode finalAggregationNode = (AggregationNode) mapEntry.getKey().getPlanNode();
+                        if (groupingCols.containsAll(finalAggregationNode.getGroupingKeys())) {
```

Comment:
Is it possible to have two aggregations with the same grouping sets hence match a different one?

Reply:
We need a better way to find which partial aggregate corresponds to which final one - I found out that there are cases where the grouping columns get renamed in the middle (perhaps by a different transformation) and we no longer can match the aggregation nodes based on grouping columns.
I'm thinking matching them either by planNode.sourceLocation or introducing another variable in aggregation nodes aggregationId to know which 2 nodes came from the same original aggregate

Reply:
I introduced a new field aggregationId in AggregationNode: it's normally empty, but when an aggregate is split into partial and final, the new nodes get the same id so they can be matched later